File: .sdk/tm/java/feature/ElementcardFeature.java
package JAVAPACKAGE.feature;

import java.util.Map;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.Result;
import JAVAPACKAGE.core.SdkClient;

// elementcard's OVERLAY for the bundled `java` target.
//
// The bundled java target's own template tree knows nothing about this
// feature and must not be edited to learn — this package ships the source
// at the layout `java` uses (`feature/<Name>Feature.java`) under its own
// `tm/java`, and the fan-out copies it across at `feature add` time.
//
// Renders a result record shaped like an element (number, symbol, name,
// mass) as an ASCII periodic-table tile. Shape-triggered, not entity-bound:
// any single-record result with those four fields gets a card. The card,
// exactly (inner width 9; number left, symbol right; name and mass centred
// with the left bias; integral numbers without a decimal point; names
// truncate at 9):
//
//   +---------+
//   |26     Fe|
//   |  Iron   |
//   | 55.845  |
//   +---------+
//
// The aggregates live on the feature as public fields (count, last) —
// java's home for the record ts keeps at `client._elementcard`, the same
// split the retry and timeout features make. `print: true` writes each
// card to stdout.
@SuppressWarnings({"unchecked"})
public class ElementcardFeature extends BaseFeature {

  private static final int CARD_WIDTH = 9;
  private static final String CARD_EDGE = "+" + "-".repeat(CARD_WIDTH) + "+";

  private SdkClient client;
  private Map<String, Object> options;

  // Activity tracking (mirrors the ts client._elementcard record).
  public int count = 0;
  public String last = "";

  public ElementcardFeature() {
    super("elementcard", "0.1.0", true);
  }

  @Override
  public void init(Context ctx, Map<String, Object> options) {
    this.client = ctx.client;
    this.options = options;
    this.active = FeatureOptions.foptBool(options, "active", false);
  }

  /** The card for one element-shaped record (mirrors the ts `render`). */
  public String render(Map<String, Object> rec) {
    return renderCard(rec);
  }

  @Override
  public void preResult(Context ctx) {
    if (!this.active) {
      return;
    }

    Result result = ctx.result;
    if (result == null) {
      return;
    }

    // preResult fires between makeResponse and makeResult, so the parsed
    // body is on the result and resdata is usually still unset; prefer
    // resdata when a feature or transform has already supplied it.
    Object data = result.resdata != null ? result.resdata : result.body;

    if (!shaped(data)) {
      return;
    }

    String card = renderCard((Map<String, Object>) data);
    this.count++;
    this.last = card;

    if (FeatureOptions.foptBool(this.options, "print", false)) {
      System.out.println(card);
    }
  }

  // True when the record is a single element-shaped result. A JSON array
  // parses to a List, never a Map, so the Map check excludes it; Boolean
  // is not a Number in java, so the numeric checks cannot admit one.
  private static boolean shaped(Object data) {
    if (!(data instanceof Map)) {
      return false;
    }
    Map<String, Object> rec = (Map<String, Object>) data;
    return rec.get("number") instanceof Number
        && rec.get("symbol") instanceof String
        && rec.get("name") instanceof String
        && rec.get("mass") instanceof Number;
  }

  // Renders a numeric value the way every target must: integral values
  // with NO decimal point ('247', '26'), anything else in its shortest
  // natural form ('55.845', '196.97'). JSON numbers arrive as Integer,
  // Long or Double depending on the parser, so normalise: integral
  // renders as long, the rest as Double.toString.
  private static String numstr(Object v) {
    Number n = (Number) v;
    if (n instanceof Double || n instanceof Float) {
      double d = n.doubleValue();
      if (!Double.isNaN(d) && !Double.isInfinite(d) && d == Math.rint(d)) {
        return Long.toString((long) d);
      }
      return Double.toString(d);
    }
    return Long.toString(n.longValue());
  }

  private static String pad(int n) {
    return n <= 0 ? "" : " ".repeat(n);
  }

  // Centres s in w with the LEFT bias (left pad = floor((w-len)/2)),
  // truncated at w.
  private static String center(String s, int w) {
    if (s.length() >= w) {
      return s.substring(0, w);
    }
    int p = w - s.length();
    int left = p / 2;
    return pad(left) + s + pad(p - left);
  }

  // The card for one element-shaped record. Lines joined with \n, no
  // trailing newline.
  private static String renderCard(Map<String, Object> rec) {
    String num = numstr(rec.get("number"));
    String sym = String.valueOf(rec.get("symbol"));
    String line2 =
        "|" + num + pad(CARD_WIDTH - num.length() - sym.length()) + sym + "|";
    return CARD_EDGE + "\n"
        + line2 + "\n"
        + "|" + center(String.valueOf(rec.get("name")), CARD_WIDTH) + "|\n"
        + "|" + center(numstr(rec.get("mass")), CARD_WIDTH) + "|\n"
        + CARD_EDGE;
  }
}

File: .sdk/tm/go/feature/elementcard_feature.go
package feature

import (
	"fmt"
	"strconv"
	"strings"

	"GOMODULE/core"
)

// elementcard's OVERLAY for the bundled `go` target.
//
// The bundled go target's own template tree knows nothing about this
// feature and must not be edited to learn — this package ships the source
// at the layout `go` uses (`feature/<name>_feature.go`) under its own
// `tm/go`, and the fan-out copies it across at `feature add` time.
//
// Renders a result record shaped like an element (number, symbol, name,
// mass) as an ASCII periodic-table tile. Shape-triggered, not entity-bound:
// any single-record result with those four fields gets a card. The card,
// exactly (inner width 9; number left, symbol right; name and mass centred
// with the left bias; integral numbers without a decimal point; names
// truncate at 9):
//
//	+---------+
//	|26     Fe|
//	|  Iron   |
//	| 55.845  |
//	+---------+
//
// The aggregates live on the feature value (Count, Last) — go's home for
// the record ts keeps at `client._elementcard`, the same split the retry
// and timeout features make. `print: true` writes each card to stdout.
type ElementcardFeature struct {
	BaseFeature
	client  *core.ElementdemoSDK
	options map[string]any

	// Activity tracking (mirrors the ts client._elementcard record).
	Count int
	Last  string
}

const elementcardWidth = 9

var elementcardEdge = "+" + strings.Repeat("-", elementcardWidth) + "+"

func NewElementcardFeature() *ElementcardFeature {
	return &ElementcardFeature{
		BaseFeature: BaseFeature{
			Version: "0.1.0",
			Name:    "elementcard",
			Active:  true,
		},
	}
}

func (f *ElementcardFeature) Init(ctx *core.Context, options map[string]any) {
	f.client = ctx.Client
	f.options = options
	f.Active = foptBool(options, "active", false)
}

// Render exposes the card for one element-shaped record (mirrors the ts
// feature's `render` method).
func (f *ElementcardFeature) Render(rec map[string]any) string {
	return elementcardRender(rec)
}

func (f *ElementcardFeature) PreResult(ctx *core.Context) {
	if !f.Active {
		return
	}

	result := ctx.Result
	if result == nil {
		return
	}

	// PreResult fires between makeResponse and makeResult, so the parsed
	// body is on the result and resdata is usually still unset; prefer
	// resdata when a feature or transform has already supplied it.
	data := result.Resdata
	if data == nil {
		data = result.Body
	}

	rec, ok := elementcardShaped(data)
	if !ok {
		return
	}

	card := elementcardRender(rec)
	f.Count++
	f.Last = card

	if foptBool(f.options, "print", false) {
		fmt.Println(card)
	}
}

// elementcardShaped reports whether the record is a single element-shaped
// result. Booleans are their own type in go, so the numeric cases below
// cannot admit one — the check the dynamic targets must make explicitly.
func elementcardShaped(data any) (map[string]any, bool) {
	rec, isMap := data.(map[string]any)
	if !isMap || rec == nil {
		return nil, false
	}
	if _, ok := elementcardNumstr(rec["number"]); !ok {
		return nil, false
	}
	if _, ok := rec["symbol"].(string); !ok {
		return nil, false
	}
	if _, ok := rec["name"].(string); !ok {
		return nil, false
	}
	if _, ok := elementcardNumstr(rec["mass"]); !ok {
		return nil, false
	}
	return rec, true
}

// elementcardNumstr renders a numeric value the way every target must:
// integral values with NO decimal point ('247', '26'), anything else in
// its shortest natural form ('55.845', '196.97'). FormatFloat's -1
// precision is exactly that rule. Non-numbers report ok=false, which is
// what the shape check keys off.
func elementcardNumstr(v any) (string, bool) {
	switch n := v.(type) {
	case int:
		return strconv.Itoa(n), true
	case int64:
		return strconv.FormatInt(n, 10), true
	case float64:
		return strconv.FormatFloat(n, 'f', -1, 64), true
	case float32:
		return strconv.FormatFloat(float64(n), 'f', -1, 32), true
	}
	return "", false
}

func elementcardPad(n int) string {
	if n <= 0 {
		return ""
	}
	return strings.Repeat(" ", n)
}

// elementcardCenter centres s in w with the LEFT bias (left pad =
// floor((w-len)/2)), truncating at w. Byte-based like the ts feature's
// char-based slice — element fields are ASCII.
func elementcardCenter(s string, w int) string {
	if len(s) >= w {
		return s[:w]
	}
	p := w - len(s)
	left := p / 2
	return elementcardPad(left) + s + elementcardPad(p-left)
}

// elementcardRender draws the card for one element-shaped record.
func elementcardRender(rec map[string]any) string {
	num, _ := elementcardNumstr(rec["number"])
	sym, _ := rec["symbol"].(string)
	name, _ := rec["name"].(string)
	mass, _ := elementcardNumstr(rec["mass"])

	line2 := "|" + num +
		elementcardPad(elementcardWidth-len(num)-len(sym)) + sym + "|"

	return elementcardEdge + "\n" +
		line2 + "\n" +
		"|" + elementcardCenter(name, elementcardWidth) + "|\n" +
		"|" + elementcardCenter(mass, elementcardWidth) + "|\n" +
		elementcardEdge
}
